Llama 4 Scout and Maverick text support

MartinGleize opened the pull request:
**What does this PR do? Please describe:**
This PR adds Llama 4 Scout and Maverick support to fairseq2.

Test plan:

No regression on Llama 3 inference recipe
Llama 4 Scout parity: verified against llama-stack on gsm8k test.
Llama 4 Scout finetuning: recipe should work but no guarantee on effectiveness (token choice routing likely needs a load balancing loss, this will be another PR).

TODO in fs2 0.5.*:
test Llama 4 Scout native inference recipe support

**Does your PR introduce any breaking changes? If yes, please list them:**
No.

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

MartinGleize commented on src/fairseq2/models/llama4/model/vision/embedding.py:
The vision stuff is mostly pure Pytorch (no re-use of fs2 components) and not properly tested (checkpoint loading and sharding was tested before 0.5, but not now). I'm in favor of keeping it still so that we can iterate on it in another PR.

MartinGleize commented on src/fairseq2/models/llama4/model/moe/token_dispatcher.py:
```diff
@@ -0,0 +1,443 @@
+    return output_splits.int().view(-1).tolist()
+
+
+class DP2EPTokenDispatcher(MoETokenDispatcher):
```
DP2EP is a specific expert parallelism mode in Llama 4, whereby experts are distributed on DP ranks (so EP gangs piggy-back onto DP gangs).

Expert parallelism is not present in this PR but I've tested it before, so I'll leave this dispatcher here for later EP support (to come in another PR).

cbalioglu commented on src/fairseq2/models/transformer/attention_bias.py:
```diff
@@ -97,6 +97,39 @@ def __repr__(self) -> str:
 
 
+@final
+class ChunkedAttentionBias(AttentionBias):
```
Have you tried out how this would play out with flash2/flash3 attention? As far as I know neither version has support for it. I also just found out this: https://x.com/winglian/status/1910430245854773523

That would mean a significant slow-down during training, especially for doc-causal masking. Looks like flex attention is the only potential short-term solution (which has another PR) that we should prioritize if we want to use LLaMA 4 for any real-world use case.

Not sure how your experience has been so far without an optimized attention support for chunked bias.

MartinGleize replied:
This is definitely an issue yes. I haven't experimented with inputs long enough for the chunk size to matter yet. But I added this issue to our internal L4 tracking doc as a priority next.

As far as I see, the reference training code heavily referenced xformers' memory-efficient attention as a backend.

cbalioglu commented on src/fairseq2/nn/grouped_projection.py:
```diff
@@ -0,0 +1,403 @@
+    def from_batch_linear(
+        linear: BatchLinear,
+        gang: Gang,
+        scatter_input: bool = False,
```
I would make the default `True` to be consistent with `ColumnShardedLinear`.

cbalioglu commented on src/fairseq2/sharder.py:
```diff
@@ -29,7 +32,20 @@
+    """Whether the sharded dimension is at the boundary between
+    two model parallel regions. Allows to avoid unnecessary communication."""
+
+    disable_end_reduce: bool = False
```
I plan to revise this API (in a BC way) so we can have more descriptive name for this. For the time being it can stay as is.

MartinGleize replied:
Reverted my changes to this and moved the custom logic to the MoESharder.

cbalioglu commented on src/fairseq2/sharder.py:
```diff
@@ -29,7 +32,20 @@
+    of the forward of a sharded layer.
+    If ``True``, it is the responsibility of the layer's user to perform the reduce."""
+
+    shard_children: bool = False
```
We likely won't need this. Let me think more about the sharding APIs. In the meantime it can stay here as it is.

cbalioglu reviewed (commented):
Sorry for the delay. It is a huge PR so took me a while to go over it. I left quite a few comments, but most of them are not really major issues and more about simplifying the code. It is clear that the core pieces of the PR were taken from the internal code base and modified for OSS, so tried to address the issues caused from that. One additional feedback is to verify that all docstrings are correct. I think overall there are inconsistencies between what is implemented and what is documented (especially in MoE, Router, and TokenDispatcher).

Thanks again for the work, looking forward to next revision!

MartinGleize commented on src/fairseq2/models/llama4/interop.py:
```diff
@@ -0,0 +1,177 @@
+    # deduce the tensor-parallelism size
+    tp_size = expected_wqkv_dim_0 // actual_wqkv_dim_0
+    # recompute local dimensions
+    n_kv_heads = config.num_key_value_heads or config.num_attn_heads
```
There were definitely redundant lines there!

MartinGleize commented on src/fairseq2/models/llama4/model/moe/router.py:
```diff
@@ -0,0 +1,130 @@
+
+    The router is never sharded.
+
+    Args:
```
Router is removed. It used to contain like 3 different implementations in earlier versions, but the few lines we have right now don't justify it.

MartinGleize commented on src/fairseq2/models/llama4/model/moe/experts.py:
```diff
@@ -0,0 +1,86 @@
+from fairseq2.nn.grouped_projection import BatchLinear, GroupedProjection
+
+
+class Experts(nn.Module):
```
Good idea. Ended up putting experts.py in fairseq2.models.transformer.
* ExpertNetwork is the ABC
* GroupedExpertNetwork is the basic unsharded implem with 3D weights and BMMs.
* TPShardedExpertNetwork is the TP-sharded version.
* We can add others when EP comes.

MartinGleize commented on src/fairseq2/models/llama4/model/moe/experts.py:
```diff
@@ -0,0 +1,86 @@
+from fairseq2.nn.grouped_projection import BatchLinear, GroupedProjection
+
+
+class Experts(nn.Module):
```
Moved experts to fairseq2.models.transformer . It was enriched, documented and made slightly more general so this felt like a good place to put it, but happy to keep it under llama4 too.

MartinGleize commented on src/fairseq2/models/llama4/model/moe/moe.py:
```diff
@@ -0,0 +1,217 @@
+        bs, slen, dim = x.shape
+        # top_scores shape (num_experts, tokens_per_expert) for EC and (bs*slen*experts_per_token, ) for TC
+        # selected_indices shape (num_experts, tokens_per_expert) for EC training,
+        # (num_experts, bs*slen) for EC inference, and (bs*slen*experts_per_token,) for TC
```
Hmm I simplified so much that I didn't realize I removed this, but I'll have another go soon at documenting the MoE forward with comments, shapes, etc to give a better idea of what's happening.

MartinGleize commented on src/fairseq2/models/llama4/model/moe/moe.py:
```diff
@@ -0,0 +1,217 @@
+        )
+
+        # Use a TP token dispatcher for non-EP scenarios
+        self.token_dispatcher = TPTokenDispatcher(num_experts)
```
Removed the dispatching, but we'll probably need to re-introduce it soon.
EP was not present in my PR (for example, it doesn't set up gangs, and it doesn't shard the expert weights, although I got both of those in our internal code) so I'll get rid of the dispatching temporarily as well.

MartinGleize commented on src/fairseq2/models/llama4/model/moe/router.py:
```diff
@@ -0,0 +1,130 @@
+            .expand(scores.size(0), -1)
+        )
+
+        scores = torch.sigmoid(scores)
```
Good question, reference code has a flag to turn this on, and llama-stack has this version. I think in the end for most models that matter, this should be done in float32.

MartinGleize commented on src/fairseq2/models/llama4/model/moe/moe.py:
```diff
@@ -0,0 +1,217 @@
+        routed_input = routed_input * top_scores.reshape(-1, 1)
+
+        # Dispatch of tokens to experts (possibly with all-to-all)
+        # permute_metadata is only useful if sharding_strategy == "dp2ep-v1"
```
There is a dp2ep-v2 :P. Anyway what we had was the v1.

MartinGleize commented on src/fairseq2/models/llama4/model/moe/moe.py:
```diff
@@ -0,0 +1,217 @@
+        # Finish shared expert computation
+        if self.shared_expert is not None:
+            assert shared_expert_out is not None
+            shared_expert_out = self.shared_expert.forward_output(shared_expert_out)
```
Yes that's it (see answer above). Honestly I'm not sure this was the product of the precise benchmarking this type of finegrained optim requires (imo). I can look into it when we introduce EP.

MartinGleize commented:
* Removed most of the currently-dead code.

* Refactored experts into a single generic experts.py including the TP-sharded impl. (in src/fairseq2/models/transformer/experts.py)

* Gathered MoE sharding logic to a single place instead of modify the core lib.

cbalioglu commented on src/fairseq2/models/transformer/experts.py:
```diff
@@ -0,0 +1,328 @@
+    Their layers are implemented as 3-D tensors used in BMM.
+    """
+
+    def _create_layers(
```
Nit: this can be a standalone function like `create_expert_glu_layers` instead of a mixin.

cbalioglu commented on src/fairseq2/models/transformer/experts.py:
Thanks a lot @MartinGleize! This looks great. Just left a very nit comment. Also note that I plan to land a change today/tomorrow about how we do model sharding that will potentially deprecate ModelSharder. I have already applied that change to other models. I can quickly do that for LLaMA-4 as well one we land this.

cbalioglu reviewed (approved):
Thanks a lot for the revision! It looks much cleaner now. I only left a very nit comment. Otherwise, looks good to me!
